Fixing errors and warnings on inventory creation

Rocco83 opened the pull request:
Characters like "." and "-" are not valid in group name
Fixes #1

Rocco83 committed:
Fixing errors and warnings on inventory creation

Characters like "." and "-" are not valid in group name

denismaggior8 commented:
Hi @Rocco83 and thanks for your contribution.
I'm reviewing your code and I can say the following:

1) About the forbidden characters, the actual syntax does not allow the usage of dots and dashes in the group names. Your code correctly strips out the forbidden chars from values coming from Shelly Cloud so they can be used when composing the group name  ✅
2) About the **cloud_online_true** group, it is ment to group all the devices that, at the time of the query, are registered to Shelly  Cloud and found actively connected. I saw the new group **cloud_online_na** you created to aggregate all the devices without `cloud: true` attribute. This is not clear to me, since in my environment the only devices without such attribute are the "virtual" devices (for example, the second channel of a two channels device is considered a virtual devices without such attribute). Wouldn't be better to just skip all the devices without `cloud: true` attribute?
3) I see that in the case a device has not an IP you rise an AnsibleError. I'm not able to test this situation but if I'm not wrong the execution would stop abnormally if a single device without IP is present in the list. Wouldn't be better to use the device mDNS name in this case?

Looking for your reply.
Denis

Rocco83 committed:
Warning is now show on the command line

Rocco83 committed:
Skipping devices without cloud_online key

As by current evidences, the devices without cloud_online are the
devices which are "virtual" (2nd channel, 3rd channel, ...)
The channel 0 device will populate properly cloud_online_true or
cloud_online_false

Rocco83 commented:
> This is not clear to me, since in my environment the only devices without such attribute are the "virtual" devices (for example, the second channel of a two channels device is considered a virtual devices without such attribute).

You are correct, the cloud_online is only on the first channel when multiple channels are existing on the device.
I personally think that this is a bug in the API, because *all* of the other values are available.

> Wouldn't be better to just skip all the devices without cloud: true attribute?

As just mentioned above, *all* of the other fiels are present, including cloud, which is mapped to true as well.
Still, as at least another json entry with a valid IP is populated with `cloud_online` `true` or `false`, 
my proposal is to skip the entry if `cloud_online` key is not populated..

Still, i think that the proper fix would be to "copy" the value from the device with channel 0, while this shall be fixed by Shelly API.

Example with a Shelly Pro 4PM
```json
{
  "key": "083af27beefc",
  "id": "083af27beefc",
  "type": "SPSW-004PE16EU",
  "category": "relay",
  "channel": 0,
  "channels_count": 4,
  "cloud_fields": {
    "cloud": true,
    "cloud_assoc": true,
    "cloud_online": true,
    "cloud_options": {
      "exclude_event_log": false
    }
  }
}
{
  "key": "083af27beefc_1",
  "id": "083af27beefc_1",
  "type": "SPSW-004PE16EU",
  "category": "relay",
  "channel": 1,
  "channels_count": 4,
  "cloud_fields": {
    "cloud": true,
    "cloud_assoc": true,
    "cloud_options": {
      "exclude_event_log": false
    }
  }
}
{
  "key": "083af27beefc_2",
  "id": "083af27beefc_2",
  "type": "SPSW-004PE16EU",
  "category": "relay",
  "channel": 2,
  "channels_count": 4,
  "cloud_fields": {
    "cloud": true,
    "cloud_assoc": true,
    "cloud_options": {
      "exclude_event_log": false
    }
  }
}
{
  "key": "083af27beefc_3",
  "id": "083af27beefc_3",
  "type": "SPSW-004PE16EU",
  "category": "relay",
  "channel": 3,
  "channels_count": 4,
  "cloud_fields": {
    "cloud": true,
    "cloud_assoc": true,
    "cloud_options": {
      "exclude_event_log": false
    }
  }
}
```

> I see that in the case a device has not an IP you rise an AnsibleError. I'm not able to test this situation but if I'm not wrong the execution would stop abnormally if a single device without IP is present in the list. 

It does not fail because there is no raise.
Still, the logging was not done properly with AnsibleError, not showing the error message.
I've updated the pushed code to show the Warning on the command line
*disclaimer* i don't know if this is the proper way to show a warning in this case, but it is working.

> Wouldn't be better to use the device mDNS name in this case?
In my case, i have no IP and no mDNS values.
As i mentioned, this is a device disconnected since a long time.
I don't know if this is the reason, but i don't see the mDNS key (case insensitive).
Unless you can propose some code directly, i would not know how to fix it, sorry...

```json
"e8db84abebcf": {
                "id": "e8db84abebcf",
                "name": "Shelly5 test",
                "image": "images/device_images/SHSW-25.png",
                "category": "roller",
                "type": "SHSW-25",
                "room_id": -10,
                "channel": 0,
                "channels_count": 2,
                "name_sync": false,
                "gen": 1,
                "mode": "roller",
                "cloud": true,
                "cloud_assoc": true,
                "template": "SHSW-25-roller",
                "position": 0,
                "modified": 1726055771,
                "cloud_options": {
                    "exclude_event_log": false
                },
                "relay_usage": "roller",
                "appliance_type": 15
            },```

denismaggior8 commented:
Ok @Rocco83 , let's proceed like this:

- Forbidden chars are now stripped out from group name: I tested your code and it seems ok
- Devices without `cloud_online` attribute are not grouped into `cloud_online_true` inventory group: I tested your code and it seems ok. We will see in the future if Shelly will fix the bug on multi-channels devices, even if I think that semantically is not correct to treat every channel as a different device with all the same attributes.
- I don't have any device without ip to test the case but it seems reasonable to me to skip the device and display a warning.

I'm going to merge this PR!
